File: backend/app/schemas/instruction.py

from app.locales import schema_field_description
from typing import Any, Dict, List, Optional, Literal
from pydantic import BaseModel, Field



class InstructionBase(BaseModel):
    op: str = Field(..., description=schema_field_description("op"))


class SetInstruction(InstructionBase):
    op: Literal["set"] = "set"
    path: str = Field(..., description=schema_field_description("path"))
    value: Any = Field(..., description=schema_field_description("value"))


class AppendInstruction(InstructionBase):
    op: Literal["append"] = "append"
    path: str = Field(..., description=schema_field_description("path"))
    value: Any = Field(..., description=schema_field_description("value"))


class DoneInstruction(InstructionBase):
    op: Literal["done"] = "done"


Instruction = SetInstruction | AppendInstruction | DoneInstruction



class GenerationConfig(BaseModel):
    mode: Literal["instruction_stream"] = Field(
        default="instruction_stream",
        description=schema_field_description("mode")
    )
    prompt_template: Optional[str] = Field(
        default=None,
        description=schema_field_description("prompt_template")
    )
    field_hints: Optional[Dict[str, str]] = Field(
        default=None,
        description=schema_field_description("field_hints")
    )
    field_order: Optional[List[str]] = Field(
        default=None,
        description=schema_field_description("field_order")
    )
    custom: Optional[Dict[str, Any]] = Field(
        default=None,
        description=schema_field_description("custom")
    )



class ConversationMessage(BaseModel):
    role: Literal["system", "user", "assistant"] = Field(..., description=schema_field_description("role"))
    content: str = Field(..., description=schema_field_description("content"))


class InstructionGenerateRequest(BaseModel):


    llm_config_id: int = Field(..., description=schema_field_description("llm_config_id"))

    user_prompt: str = Field(default="", description=schema_field_description("user_prompt"))

    response_model_schema: Dict[str, Any] = Field(..., description=schema_field_description("response_model_schema"))

    current_data: Dict[str, Any] = Field(default_factory=dict, description=schema_field_description("current_data"))

    conversation_context: List[ConversationMessage] = Field(
        default_factory=list,
        description=schema_field_description("conversation_context")
    )

    generation_config: Optional[GenerationConfig] = Field(
        default=None,
        description=schema_field_description("generation_config")
    )

    prompt_template: Optional[str] = Field(
        default=None,
        description=schema_field_description("prompt_template")
    )

    context_info: Optional[str] = Field(
        default=None,
        description=schema_field_description("context_info")
    )

    temperature: Optional[float] = Field(default=0.7, description=schema_field_description("temperature"))
    max_tokens: Optional[int] = Field(default=None, description=schema_field_description("max_tokens"))
    timeout: Optional[float] = Field(default=150, description=schema_field_description("timeout"))

    deps: Optional[str] = Field(default=None, description=schema_field_description("deps"))



class ThinkingEvent(BaseModel):
    type: Literal["thinking"] = "thinking"
    text: str = Field(..., description=schema_field_description("text"))


class InstructionEvent(BaseModel):
    type: Literal["instruction"] = "instruction"
    instruction: Instruction = Field(..., description=schema_field_description("instruction"))


class WarningEvent(BaseModel):
    type: Literal["warning"] = "warning"
    text: str = Field(..., description=schema_field_description("text"))


class ErrorEvent(BaseModel):
    type: Literal["error"] = "error"
    text: str = Field(..., description=schema_field_description("text"))


class DoneEvent(BaseModel):
    type: Literal["done"] = "done"
    success: bool = Field(default=True, description=schema_field_description("success"))
    message: Optional[str] = Field(default=None, description=schema_field_description("message"))


StreamEvent = ThinkingEvent | InstructionEvent | WarningEvent | ErrorEvent | DoneEvent

File: frontend/src/renderer/src/types/instruction.ts


export type InstructionOp = 'set' | 'append' | 'done'

export interface InstructionBase {
  op: InstructionOp
}

export interface SetInstruction extends InstructionBase {
  op: 'set'
  path: string
  value: any
}

export interface AppendInstruction extends InstructionBase {
  op: 'append'
  path: string
  value: any
}

export interface DoneInstruction extends InstructionBase {
  op: 'done'
}

export type Instruction = SetInstruction | AppendInstruction | DoneInstruction


export interface GenerationConfig {
  mode?: 'instruction_stream'
  prompt_template?: string
  field_hints?: Record<string, string>
  field_order?: string[]
  custom?: Record<string, any>
}


export interface ConversationMessage {
  role: 'system' | 'user' | 'assistant'
  content: string
}

export interface InstructionGenerateRequest {
  llm_config_id: number
  user_prompt?: string
  response_model_schema: Record<string, any>
  current_data?: Record<string, any>
  conversation_context?: ConversationMessage[]
  generation_config?: GenerationConfig
  prompt_template?: string
  context_info?: string
  temperature?: number
  max_tokens?: number
  timeout?: number
  deps?: string
}


export interface ThinkingEvent {
  type: 'thinking'
  text: string
}

export interface InstructionEvent {
  type: 'instruction'
  instruction: Instruction
}

export interface WarningEvent {
  type: 'warning'
  text: string
}

export interface ErrorEvent {
  type: 'error'
  text: string
}

export interface DoneEvent {
  type: 'done'
  success?: boolean
  message?: string
}

export type StreamEvent = ThinkingEvent | InstructionEvent | WarningEvent | ErrorEvent | DoneEvent


export type GenerationMessageType = 'thinking' | 'action' | 'system' | 'user' | 'warning' | 'error'

export interface GenerationMessage {
  type: GenerationMessageType
  content: string
  timestamp: number
}
